onnxruntime/core/graph/graph_utils.cc

```diff
@@ -191,51 +193,121 @@ static void UpdateImplicitInputNameInSubgraph(Node& node,
+  std::vector<GraphEdge> output_edges = RemoveAndReturnOutputEdges(graph, node);
 
-    Node& mutable_output_edge_node = *graph.GetNode(output_edge_node_index);
+  // Store info for the input edge.
+  const Node::EdgeEnd& input_edge_end = *node.InputEdgesBegin();
+  const auto& input_edge_node = input_edge_end.GetNode();
+  const GraphEdge input_edge{input_edge_node.Index(),
+                             node.Index(),
+                             input_edge_end.GetSrcArgIndex(),
+                             input_edge_end.GetDstArgIndex(),
+                             GetNodeOutputName(input_edge_node, input_edge_end.GetSrcArgIndex())};
 
-    if (CanUpdateImplicitInputNameInSubgraph(mutable_output_edge_node, removed_output_name, output_name)) {
-      UpdateImplicitInputNameInSubgraph(mutable_output_edge_node, removed_output_name, output_name);
-    } else {
-      // we can't safely remove this node
+  // Remove node (this will remove the input edge too).
+  graph.RemoveNode(node.Index());
+
+  // Create connections between the incoming node and the outgoing nodes of the node that we removed.
+  for (const auto& output_edge : output_edges) {
+    // Take care of subgraph inputs.
+    if (!CheckAndUpdateImplicitInputInSubgraph(graph, input_edge.arg_name, output_edge)) {
       return false;
     }
+
+    // Add new edge connecting the input with the output nodes directly.
+    graph.AddEdge(input_edge.src_node, output_edge.dst_node, input_edge.src_arg_index, output_edge.dst_arg_index);
   }
 
-  // Remove output edge.
-  graph.RemoveEdge(node.Index(), output_edge_node_index, output_edge.GetSrcArgIndex(), output_edge_dst_arg);
+  return true;
+}
 
-  // Remove node (this will remove the input edge too).
+/** Remove a node with a single incoming initializer (and no other incoming node). */
+static bool RemoveNodeWithSingleInitializerIn(Graph& graph, Node& node) {
+  // Store info for and remove the output edges.
+  std::vector<GraphEdge> output_edges = RemoveAndReturnOutputEdges(graph, node);
+
+  // Get incoming initializer node arg.
+  NodeArg* input_def = node.MutableInputDefs()[0];
+
+  // Remove the node.
   graph.RemoveNode(node.Index());
 
-  // Add new edge connecting the input with the output nodes directly.
-  graph.AddEdge(input_edge_node_index, output_edge_node_index, input_edge_dst_arg, output_edge_dst_arg);
+  // Add the incoming initializer as input to the outgoing nodes of the node that we removed.
+  for (auto& output_edge : output_edges) {
+    // Take care of subgraph inputs.
+    if (!CheckAndUpdateImplicitInputInSubgraph(graph, input_def->Name(), output_edge)) {
+      return false;
+    }
+
+    // Replace outgoing node's input to use the initializer.
+    auto output_node = graph.GetNode(output_edge.dst_node);
+    if (!output_node) {
+      return false;
+    }
+
+    auto& dst_def = output_node->MutableInputDefs()[output_edge.dst_arg_index];
+    dst_def = input_def;
+  }
```

Comment:
nit: slightly more readable if you just do 
  output_node->MutableInputDefs()[output_edge.dst_arg_index] = input_def;